File: github_cpp/19/15.c
#include<iostream>
using namespace std;

void printArray(int *input, int size);
void qsort(int *input,int low, int high);
int partition(int *input, int low, int high);
void swap(int *a, int *b);
int main() {
	int input[] = {13,19,9,5,12,8,7,4,21,2,6,11};
	int size = sizeof(input) / sizeof(input[0]);

	printArray(input,size);
	qsort(input,0,size-1);
	printArray(input,size);

	getchar();
	return 0;
}
int partition(int *input, int low, int high) {
	int i = low;
	int x = input[high];

	for (auto j = low; j < high; j++) {
		if (input[j]<x) {
			swap(&input[i], &input[j]);
			i++;
		}
	}
	swap(&input[i], &input[high]);
	return i;
}
void qsort(int *input, int low, int high) {
	if (low>=high) {
		return;
	}
	int pivot = partition(input, low, high);
	qsort(input, low, pivot-1);
	qsort(input, pivot+1, high);
}
void swap(int *a, int *b) {
	int temp = *a;
	*a = *b;
	*b = temp;
}
void printArray(int *input, int size) {
	for (auto i = 0; i < size; i++) {
		cout << input[i] << " ";
	}
	cout << endl;
}

File: github_java_sort/8/15.java

public class Quicksort implements Runnable{
	int[] arr;
	int low;
	int high;
	
	public Quicksort(int[] arr,int low,int high) {
		
		this.arr = arr;
		this.low= low;
		this.high = high;
	}
	
	void QS() throws InterruptedException{
		if(low<high){
			int p = partition();
			Quicksort obj1=new Quicksort(arr,low,p);
			Thread t1=new Thread(obj1);
			t1.start();
			
			
			Quicksort obj2=new Quicksort(arr,p+1,high);
			Thread t2=new Thread(obj2);
			t2.start();
			
			t1.join();
			t2.join();
		}
	}
	
	int partition(){
		int pivot = arr[low];
		int i = low-1;
		int j = high+1;
		
		while(true){
			do{
				i = i+1;
			}while(arr[i]<pivot);
			
			do{
				j = j-1;
			}while(arr[j]>pivot);
			
			if(i>=j){
				return j;
			}
			
			int temp = arr[j];
			arr[j] = arr[i];
			arr[i] = temp;
		}
		
	}

	@Override
	public void run() {
		
		try {
			QS();
		} catch (InterruptedException e) {
			
			e.printStackTrace();
		}	
	}

}
